### ISSUE TITLE
Tracking Issue for rustdoc `--doctest-compilation-args` command line option

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#134172]
### Steps

[Original issue](https://github.com/rust-lang/rust/issues/67533)

- [ ] [Implementation PR](https://github.com/rust-lang/rust/pull/128780)
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr

**Comment by weihanglo** [comment]
Considering all linked issues (#67533, #66623, rust-lang/cargo#13697, rust-lang/miri#3869) are actually Cargo-related, I wonder how to integrate this flag to help Cargo users. Cargo doesn't parse `RUSTFLAGS` so it may not know if `--cap-lints=warn` is there and pass to `--doctest-compilation-args` accordingly.

Also, the feature seems to open a door way bigger than what #67533 was looking for. It accepts arbitrary rustc flags. Mind sharing why propagating `-cap-lints` isn't under consideration? Maybe there are other use cases for arbitrary flags I am missing. We all know that once we provide more stuff to users, we might have no chance to take it back 😓.


That said, I don't think this feature on its own is bad. Thank you for working on a fix!!

**Comment by josephlr** [comment]
I can offer some context on why this feature might be useful if properly integrated into `cargo`. In [`getrandom`'s tests](https://github.com/rust-random/getrandom/blob/9b78fcc83ad54d2a0e045090d6fa90b2a4b315c1/.github/workflows/tests.yml#L160-L164), we want to run our existing tests with the unstable `-Zsanitizer=memory` flag.

If we replace instances of:
```bash
cargo test --target=x86_64-unknown-linux-gnu
```
with
```bash
RUSTFLAGS="-Zsanitizer=memory" cargo test -Zbuild-std --target=x86_64-unknown-linux-gnu
```
our building, linking, and testing work exactly as expected, **except for doctests**.

For doctests, as far as I can tell, it's not currently possible to pass `-Zsanitizer=memory` in all the necessary locations to make doctests work. Specifically, even the following does not work (on the currently nightly):
```bash
RUSTDOCFLAGS="-Zsanitizer=memory --doctest-build-arg=-Zsanitizer=memory" RUSTFLAGS="-Zsanitizer=memory" cargo test --doc -Zbuild-std --target=x86_64-unknown-linux-gn
```
giving the following error:
```
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.03s
   Doc-tests getrandom
error: mixing `-Zsanitizer` will cause an ABI mismatch in crate `getrandom`
  |
  = help: the `-Zsanitizer` flag modifies the ABI so Rust crates compiled with different values of this flag cannot be used together safely
  = note: unset `-Zsanitizer` in this crate is incompatible with `-Zsanitizer=memory` in dependency `core`
  = help: set `-Zsanitizer=memory` in this crate or unset `-Zsanitizer` in `core`
  = help: if you are sure this will not cause problems, you may use `-Cunsafe-allow-abi-mismatch=sanitizer` to silence this error
```

While we can just disable doctests when testing with the sanitizer, I think this provides a good test case for making sure the correct build flags are passed everywhere.

It often seems like when doing anything complicated with `RUSTFLAGS` or testing in general, the answer is "just don't use doctests". However, I really like doctests, and would like to see them well integrated with the rest of the Rust testing story.

In an ideal world, setting `RUSTFLAGS` should work for all build invocations, including those for doctests, and `RUSTDOCFLAGS`/`--doctest-build-arg` wouldn't even be necessary. However, I understand that doing this in a backwards-compatible way might be challenging.

**Comment by fmease** [comment]
Thanks for your report. However, I'd like to draw attention to the prominent banner in the issue description, most notably this part:

> A tracking issue is however *not* meant for large scale discussion, questions, or bug reports.
> Instead, open a dedicated issue for the specific matter or post in the [#t-rustdoc](https://rust-lang.zulipchat.com/#narrow/channel/266220-t-rustdoc) Zulip channel.

Moreover, it's unclear how your issue relates to the experimental command-line option `--doctest-build-arg` which this issue is tracking, I see no mention of it in your report.

Regarding your following statement:

> My expectation would be that even if rustdoc ignores `RUSTFLAGS`, it would do so consistently. Right now, you can apparently end up in a situation where a rustdoc snippet is compiled using `RUSTDOCFLAGS`, but the *modules* that it imports are compiled using `RUSTFLAGS`.

You are mixing up two different concepts, namely the build system Cargo and the programs rustdoc / rustc. rustdoc doesn't know anything about `RUSTDOCFLAGS` or `RUSTFLAGS` and rustc doesn't know anything about `RUSTFLAGS`, these environment variables are only read & understood by Cargo.

> It's not hard to imagine how this could lead to unsoundness.

It might indeed lead to *library* unsoundness (not *language* unsoundness) if the assumptions made by the library author turn out to be false (either due to a Cargo / rustdoc bug or due to misconceptions of the author).

---

It would be great if you could create a separate issue with a *minimal reproducer* so it's easier for us to see what might be going on. https://github.com/rust-lang/rust/issues is fine, I'll transfer to https://github.com/rust-lang/cargo/issues if needed, once I can visualize the problem since right now I can only speculate that some flags might not get forwarded from rustdoc to the rustc used for compiling doctests or that inlined cross-crate re-exports could play a role here (where rustdoc inlines items defined in a crate that was compiled with rustc instead of rustdoc).

**Comment by valadaptive** [comment]
I've removed the original comment. I don't think it needs much discussion and is probably intended behavior, albeit confusing. I think opening an issue would get it closed as a duplicate of this one, so I'll just wait for the broader "passing RUSTFLAGS into rustdoc" problem (is there a tracking issue for that?) to be solved.

The closest thing I found on Google relating to my issue was https://internals.rust-lang.org/t/can-we-do-something-about-rustflags-rustdocflags/21513, which referenced #67533. That issue was closed by https://github.com/rust-lang/rust/pull/128780, which is being tracked in this issue. That led me to assume #67533 is "fixed" and you won't accept any more discussion on *that* issue, so I couldn't really find an ideal place to bring it up.

After thinking things over, I think I was "reporting" a known issue, albeit in a roundabout way, and we're just waiting on someone to resolve the open questions and stabilize things. Apologies for wasting your time.
